tracer/src/Datadog.Trace/Telemetry/TelemetryControllerV2.cs

```diff
@@ -289,10 +283,139 @@ public enum ItemType
+            ShouldFlushTelemetry = false; // wait for initialization before flushing metrics
+            _lastFlush = _lastAggregation = _clock.UtcNow;
+
+            // Using a task array instead of overloads to avoid allocating the array every loop
+            _tasks = new Task[3];
+            _tasks[1] = processExitSource.Task;
+            _tasks[2] = _tracerInitialized.Task;
+        }
+
+        public interface IDelayFactory
+        {
+            Task Delay(TimeSpan delay);
+        }
+
+        public bool ShouldAggregateMetrics { get; private set; }
+
+        public bool ShouldFlushTelemetry { get; private set; }
+
+        public void SetFlushInterval(TimeSpan flushInterval)
+        {
+            _flushInterval = flushInterval;
+        }
+
+        public void SetTracerInitialized()
+        {
+            _tracerInitialized.TrySetResult(true);
+        }
+
+        public async Task WaitForNextInterval()
+        {
+            // Calculate how long before the next flush. Accounts for the fact that it might
+            // take a long time to push telemetry if the network is slow or faulty
+            var nextAggregation = _lastAggregation.Add(_metricsAggregationInterval);
+
+            // Note that we don't start flushing until initialized
+            var nextFlush = _lastFlush.Add(_flushInterval);
+            var nextAction = !_initializationFlushExecuted || nextAggregation < nextFlush ? nextAggregation : nextFlush;
+            var waitPeriod = nextAction - _clock.UtcNow;
+
+            Task? completedTask = null;
+            if (waitPeriod <= TimeSpan.Zero)
+            {
+                Log.Debug(
+                    "Time to push telemetry exceeded the flush interval, triggering the next iteration immediately");
+            }
+            else
+            {
+                _tasks[0] = _delayFactory.Delay(waitPeriod);
+                completedTask = await Task.WhenAny(_tasks).ConfigureAwait(false);
+            }
+
+            if (_processExitSource.Task.IsCompleted)
+            {
+                // end of the line, flush everything, don't bother recalculating;
+                ShouldAggregateMetrics = true;
+                ShouldFlushTelemetry = true;
+                return;
+            }
+
+            // Should we aggregate metrics?
+            var now = _clock.UtcNow;
+            ShouldAggregateMetrics = nextAggregation <= now;
+            if (ShouldAggregateMetrics)
+            {
+                _lastAggregation = now;
+            }
+
+            // Should we flush telemetry?
+            if (completedTask == _tracerInitialized.Task)
+            {
+                _initializationFlushExecuted = true;
+                // We've just been started, so should always flush telemetry
+                ShouldFlushTelemetry = true;
+                // replace the tracerInitializedTask with a task that never completes
+                _tasks[2] = Task.Delay(Timeout.Infinite);
```

Comment:
I don't like that the code depends on the index of the array even though it's allocated outside of the function. Can you move the indexes to constants to make sure both parts of the code are kept in sync?

Reply:
Done in https://github.com/DataDog/dd-trace-dotnet/pull/4388/commits/2a2bea43bdafe8ad80c677bbc85e90f76d730b7f